### PR TITLE
PROV: Add a DER to RSA-PSS deserializer implementation

### PR DESCRIPTION
**Pull request by levitte**
This is made on top of #12410 and will stay WIP until that one is merged.

### CODE REVIEW

File: crypto/serializer/deserializer_lib.c [hunk 1]
```diff
@@ -181,7 +181,7 @@ int OSSL_DESERIALIZER_CTX_add_extra(OSSL_DESERIALIZER_CTX *ctx,
     size_t w_prev_start, w_prev_end; /* "previous" deserializers */
     size_t w_new_start, w_new_end;   /* "new" deserializers */
     size_t count; /* Used to calculate how many were added in each iteration */
```
Comment (slontis) [inline crypto/serializer/deserializer_lib.c:183]: Looks like you should also set count to zero here..

You have a continue inside the do loop (which will then go to the while check with count uninited

Comment (levitte) [reply crypto/serializer/deserializer_lib.c:183]: Uhmmm, the continues are inside the for loop that is inside the do loop, no?

Comment (slontis) [reply crypto/serializer/deserializer_lib.c:183]: yep you are right..

File: test/serdes_test.c [hunk 2]
```diff
@@ -0,0 +1,454 @@
+/*
```
Comment (slontis) [inline test/serdes_test.c:1]: Does this file need quite a few
```
#ifndef OPENSSL_NO_RSA
```

Comment (levitte) [reply test/serdes_test.c:1]: No.  RSA is not disablable.

The `OPENSSL_NO_RSA` that you see sprinkled in the code are remnants from before 1.1.0, and should really be cleaned away.

Comment (slontis) [reply test/serdes_test.c:1]: Ok .. I think I have added some new ones in 3.0 then.

Comment (levitte) [reply test/serdes_test.c:1]: (pre-1.1.0 OpenSSL was much wilder...  you could try to disabled *anything*, whether it made sense or not.  Also, I'm sure you remember the time when using the RSA algorithm was legally dodgy, so people had much more reason to disable it at the time than now)

**Comment by levitte** [comment]
Now that #12410 has been merged, this got rebased and squashed, and is now out of WIP

File: providers/implementations/serializers/deserialize_der2rsa.c [hunk 3]
```diff
@@ -37,10 +39,12 @@ static OSSL_FUNC_deserializer_export_object_fn der2rsa_export_object;
 struct der2rsa_ctx_st {
     PROV_CTX *provctx;
 
+    int type;
+
     struct pkcs8_encrypt_ctx_st sc;
 };
 
-static void *der2rsa_newctx(void *provctx)
+static void *der2rsa_newctx_int(void *provctx)
```
Comment (slontis) [inline providers/implementations/serializers/deserialize_der2rsa.c:47]: I would have passed the type in here also.. but this is ok.

File: providers/implementations/serializers/deserialize_der2rsa.c [hunk 4]
```diff
@@ -177,10 +199,27 @@ static int der2rsa_deserialize(void *vctx, OSSL_CORE_BIO *cin,
...
+
+        switch (RSA_test_flags(rsa, RSA_FLAG_TYPE_MASK)) {
+        case RSA_FLAG_TYPE_RSA:
+            object_type = "RSA";
+            break;
+        case RSA_FLAG_TYPE_RSASSAPSS:
+            object_type = "RSA-PSS";
+            break;
+        default:
+            ERR_raise_data(ERR_LIB_PROV, PROV_R_INVALID_RSA_KEY,
```
Comment (slontis) [inline providers/implementations/serializers/deserialize_der2rsa.c:212]: Is there a reason why this is the only source file in providers that uses ERR_raise_data

File: test/serdes_test.c [hunk 5]
```diff
@@ -93,6 +95,14 @@ static int test_serialize_deserialize(EVP_PKEY *pkey,
         && !TEST_ptr(EVP_PKEY_get0(pkey2)))
         goto end;
 
+    /*
+     * TODO(3.0) Remov this when PEM_write_bio_PrivateKey_traditional()
```
Comment (slontis) [inline test/serdes_test.c:99]: Remove

File: providers/implementations/serializers/deserialize_der2rsa.c (hunk 4 above)
Comment (levitte) [reply providers/implementations/serializers/deserialize_der2rsa.c:212]: That's not true:

``` console
$ git grep -l ERR_raise_data providers
providers/implementations/rands/seeding/rand_vms.c
providers/implementations/signature/rsa.c
```

I guess that we all haven't found reason to add extra data that much, or simply haven't gotten used to it (the older `ERR_add_error_data` is klunky and quite discouraging to use!).

We do have an outstanding job to replace all the FOOerr() with ERR_raise(), and FOOerr() + ERR_add_error() with ERR_raise_data().  There are quite a few, but there is a script `util/err-to-raise` that's doing much of the job (thank you @richsalz) and that should be unleashed at some point.

**Comment by levitte** [comment]
> A bit quicker to review than the last one :)

Just you wait for #12512 to be done :wink:

File: providers/implementations/serializers/deserialize_der2rsa.c (hunk 4 above)
Comment (slontis) [reply providers/implementations/serializers/deserialize_der2rsa.c:212]: ProvErr() seems simpler to me.. oh well.

**Comment by slontis** [comment]
> > A bit quicker to review than the last one :)
> 
> Just you wait for #12512 to be done 😉

Well it has 64 files already... cant wait..

File: providers/implementations/serializers/deserialize_der2rsa.c (hunk 4 above)
Comment (levitte) [reply providers/implementations/serializers/deserialize_der2rsa.c:212]: The combination FOOerr() + ERR_add_error() seems simpler to you?????  I find it yucky.  But ok...

File: providers/implementations/serializers/deserialize_der2rsa.c (hunk 3 above)
Comment (levitte) [reply providers/implementations/serializers/deserialize_der2rsa.c:47]: Done

File: test/serdes_test.c (hunk 5 above)
Comment (levitte) [reply test/serdes_test.c:99]: Done

File: test/serdes_test.c [hunk 6]
```diff
@@ -95,14 +95,6 @@ static int test_serialize_deserialize(const char *type, EVP_PKEY *pkey,
         && !TEST_ptr(EVP_PKEY_get0(pkey2)))
         goto end;
 
-    /*
-     * TODO(3.0) Remov this when PEM_write_bio_PrivateKey_traditional()
-     * handles provider side keys.
-     */
-    if (make_legacy
```
Comment (slontis) [inline test/serdes_test.c]: Did you mean to remove the code also?

Comment (levitte) [reply test/serdes_test.c]: Yes, it was a duplicate of the same lines above...  that probably happened when I rebased this after #12410 was merged

**Comment by levitte** [comment]
> LGTM if the tests pass.

Looks that way

**Comment by levitte** [comment]
Merged

456b3b97a489d4be42f4258cb65bf76dfd8bab00 EVP, PROV: Add misc missing bits for RSA-PSS
a4e55cccc9991b35c3f4b3f4aac32b59aa598584 PROV: Add a DER to RSA-PSS deserializer implementation
846f96f821260ca83cc93bfa35207864b05abec5 TEST: Add RSA-PSS cases in test/serdes_test.c

**Pull request closed without merging**